plugins/ingest/src/main/java/org/elasticsearch/plugin/ingest/transport/simulate/SimulatedItemResponse.java

```diff
@@ -0,0 +1,169 @@
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.elasticsearch.plugin.ingest.transport.simulate;
+
+import org.elasticsearch.ExceptionsHelper;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.io.stream.Streamable;
+import org.elasticsearch.common.xcontent.ToXContent;
+import org.elasticsearch.common.xcontent.XContentBuilder;
+import org.elasticsearch.common.xcontent.XContentBuilderString;
+import org.elasticsearch.ingest.Data;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+
+public class SimulatedItemResponse implements Streamable, ToXContent {
+
+    private Data data;
+    private List<ProcessorResult> processorResultList;
+    private Throwable failure;
+
+    public SimulatedItemResponse() {
+
+    }
+
+    public SimulatedItemResponse(Data data) {
+        this.data = data;
+    }
+
+    public SimulatedItemResponse(List<ProcessorResult> processorResultList) {
+        this.processorResultList = processorResultList;
+    }
+
+    public SimulatedItemResponse(Throwable failure) {
+        this.failure = failure;
+    }
+
+    public boolean isFailed() {
+        if (failure != null) {
+            return true;
+        } else if (processorResultList != null) {
+            for (ProcessorResult result : processorResultList) {
+                if (result.isFailed()) {
+                    return true;
+                }
+            }
+        }
+
+        return false;
+    }
+
+    public boolean isVerbose() {
+        return this.processorResultList != null;
+    }
+
+    public Data getData() {
+        return data;
+    }
+
+    public List<ProcessorResult> getProcessorResultList() {
+        return processorResultList;
+    }
+
+    @Override
+    public void readFrom(StreamInput in) throws IOException {
+        boolean isFailed = in.readBoolean();
+        boolean isVerbose = in.readBoolean();
+        if (isFailed) {
+            this.failure = in.readThrowable();
+        } else if (isVerbose) {
+            int size = in.readVInt();
+            processorResultList = new ArrayList<>();
+            for (int i = 0; i < size; i++) {
+                ProcessorResult processorResult = new ProcessorResult();
+                processorResult.readFrom(in);
+                processorResultList.add(processorResult);
+            }
+        } else {
+            String index = in.readString();
```

Comment:
scratch that, we can't make Data implement Streamable. WE could have either static helper methods or an enhanced Data class for the plugin. I have a small preference for the latter that sounds a bit more elegant.